Add AMI comparisons to postprocessing

mdahlhausen opened the pull request:
Pull request overview
---------------------

This PR adds AMI data comparisons from EULP to the postprocessing workflow. The AMI data is available on AWS in an Athena table.

This includes:
- methods to download and process timeseries AMI data
- methods to download and process timeseries ComStock data
- adds an `ami_comparison.csv` to create a 10k `buildstock.csv` sample to compare against the AMI regions
- adds plotting for stacked end use plots by season and day type, either by kwh per sf, or annual or daytype normalizations
- adds plotting for load duration curves
- adds a test

### Pull Request Author

This pull request makes changes to (select all the apply):
 - [x] Postprocessing

Author pull request checklist:
 - [x] Tagged the pull request with the appropriate label (documentation, infrastructure, sampling, workflow measure, upgrade measure, reporting measure, postprocessing) to help categorize changes in the release notes.
 - [x] Added tests for postprocessing
 - [x] 10k+ test run
 - [ ] ComStock documentation updated
 - [ ] Changes reflected in `README.md` files
 - [x] Added 'See ComStock License' language to first two lines of each code file
 - [ ] All new and existing tests pass the CI

### Specific Review Requests
- What, if any content should I add to the postprocessing README?
- Should some of AMI comparisons be included in the release or documentation? Perhaps a separate calibration report to accompany the reference documentation?
- I don't often write in python, so my syntax may be messy. code review it please.
- What other plots should I include?

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a code review on GitHub
 - [ ] All related changes have been implemented: data and method additions, changes, tests
 - [ ] If fixing a defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] Reviewed change documentation
 - [ ] Ensured code files contain License reference
 - [ ] Results differences are reasonable
 - [ ] CI status: all tests pass

mdahlhausen committed:
add AMI data download

mdahlhausen committed:
Merge branch 'eia_comparisons' into ami_comparisons

mdahlhausen committed:
add ami aggregations

mdahlhausen committed:
add building type weights to comstock data object

mdahlhausen committed:
download total timeseries by region

mdahlhausen committed:
add comstock enduse timeseries download

mdahlhausen committed:
timeseries to long format

mdahlhausen committed:
add size calculations

mdahlhausen committed:
Create ami_comparison.csv

To generate a buildstock.csv to compare against ami data

mdahlhausen committed:
update test names

mdahlhausen committed:
save timeseries in long format with kwh per sf

mdahlhausen committed:
make timeseries data one object

mdahlhausen committed:
Merge branch 'main' into ami_comparisons

mdahlhausen committed:
align ami timeseries data names and join datasets

mdahlhausen committed:
Merge branch 'eia_comparisons' into ami_comparisons

mdahlhausen committed:
add ComStock license to new files

mdahlhausen committed:
initial plotting code

mdahlhausen committed:
Merge branch 'main' into ami_comparisons

mdahlhausen committed:
changes for initial plots

mdahlhausen committed:
add load duration curve

mdahlhausen committed:
data checks and load duration curve uncertainty

- add checks to make sure comstock and ami data are present
- use fixed max uncertainty for load duration curve

mdahlhausen committed:
Merge branch 'main' into ami_comparisons

mdahlhausen committed:
fix ploting, enduses, year-daytype, and comment

Fixes several issues:
- repeated plotting of uncertainty on end use charts
- data length mismatch
- weekday/weekends not matched between ami and comstock because of different years
- add normalization by annual total and daytype
- add function comment headers

mdahlhausen committed:
Update test_comstock_to_ami_comparison.py

set reloads to False so test runs fully

mdahlhausen commented:
There are some messy warning about enumerations when I run the test. Those seem to come from other parts of the code. Perhaps worth making a github issue if you can track it down.

mdahlhausen committed:
organize AMI graphics by building type

mdahlhausen committed:
add comstock building count to AMI plots

mdahlhausen committed:
remove extraneous arguments in ami make_plots()

mdahlhausen committed:
skip graphics if fewer than 5 AMI buildings

mdahlhausen committed:
collect stacked end use plot data for error analysis

eringold committed:
add top level import

asparke2 committed:
Corrects S3 location for AMI test run

asparke2 committed:
Updates postprocessing for compatibility with latest polars and buildstock-query

eringold committed:
Merge branch 'ami_comparisons' of https://github.com/NREL/ComStock into ami_comparisons

eringold committed:
fix for timestamp keyerror

eringold committed:
add log level to test

eringold committed:
set index and parse dates when loading from csv

eringold committed:
add comstockpostproc script

eringold commented on postprocessing/compare_comstock_to_ami.py:
Adding ability to run as a script, simliar to other 'compare_comstock_to' scripts

eringold commented on postprocessing/comstockpostproc/ami.py:
```diff
@@ -0,0 +1,257 @@
+                 raise FileNotFoundError(
+                    f'Cannot find {file_path} to reload data, set reload_from_csv=False to create CSV.')
+            logger.info(f'Reloading from CSV: {file_path}')
+            self.ami_timeseries_data = pd.read_csv(file_path, low_memory=False, index_col='timestamp', parse_dates=True)
```
needed to set index col and dtype when loading from csv

eringold commented on postprocessing/comstockpostproc/ami.py:
```diff
@@ -0,0 +1,257 @@
+                    logger.warning(f'ami data for ' + file_name + ' does not exist.')
+                    continue
+
+                df = pd.read_csv(file_path, parse_dates=['timestamp'], index_col='timestamp')
```
index col and dtype set when not reloading from csv

eringold commented on postprocessing/comstockpostproc/comstock.py:
```diff
@@ -229,6 +237,134 @@ def download_data(self):
+                 raise FileNotFoundError(
+                    f'Cannot find {file_path} to reload data, set reload_from_csv=False to create CSV.')
+            logger.info(f'Reloading from CSV: {file_path}')
+            self.ami_timeseries_data = pd.read_csv(file_path, low_memory=False, index_col='timestamp', parse_dates=True)
```
set index col and dtype when loading from csv

eringold commented on postprocessing/comstockpostproc/comstock.py:
```diff
@@ -229,6 +237,134 @@ def download_data(self):
+        agg_df['day'] = agg_df.index.day
+        agg_df['hour'] = agg_df.index.hour
+        agg_df = agg_df.groupby(['building_type', 'year', 'month', 'day', 'hour']).sum().reset_index()
+        agg_df['timestamp'] = agg_df.apply(
```
timestamp col as datetime

eringold commented on postprocessing/comstockpostproc/comstock.py:
```diff
@@ -229,6 +237,134 @@ def download_data(self):
+            value_vars=val_vars,
+            var_name='enduse',
+            value_name='kwh'
+        ).set_index('timestamp')
```
timestamp as index

eringold reviewed (dismissed):
Plots generate when downloading or reloading data. Looks good!

eringold commented:
@mdahlhausen Please see if everything still works on your end. If so, I think it's good to merge.

asparke2 reviewed (changes requested):
**What, if any content should I add to the postprocessing README?**
For the README, write a little section like [this](https://github.com/NREL/ComStock/blob/ami_comparisons/postprocessing/README.md#comparing-one-or-more-comstock-runs-to-each-other-and-cbecs) that explains how to do an AMI comparison. @eringold committed a template script already, so just describe how to use it.

**Existing Plots and Code**
The plots now work for me as well.  Some of them have a slightly clipped peak (see example below), but I don't know that it's important enough to bother fixing unless it's something quick. Otherwise I think the code is fine.

![image](https://github.com/NREL/ComStock/assets/2009290/e95dfdf7-0d7b-49ef-95fe-4587ea53c86f)

**Other Plots**
Two other plots that could be useful:
1. A comparison of all the AMI for each building type plotted on one plot. Could do it normalized and absolute. Basically to give people an understanding of how similar/not the AMI is from region to region.
2. The plot that showed an example week of data for both AMI and ComStock. Sometimes it's easier to pick out things like a time shift when you can see the whole week at once.

**Where to Release Comparisons**
I think we'd like to release the comparisons with the next EUSS release. The first version might just be a folder with a bunch of images on OEDI. We could also try to make a LaTeX template document that just has the hard-coded paths in it and generate a PDF with just a bunch of figures and little/no discussion.

mdahlhausen commented:
Thanks @eringold and @asparke2.

I'll address two items before merging:
- Adding README content
- The clipped y-axis on some plots

I've been working on summary graphic and statistics that we can use in reports and to guide calibration at a high level. I'll add those in a follow-up PR, along with the two other requested graphs (a plot comparing AMI data, week realizations).

mdahlhausen committed:
add  README

mdahlhausen committed:
add missing argument to cbecs test

mdahlhausen committed:
Merge branch 'main' into ami_comparisons

mdahlhausen committed:
increase max pool connections

asparke2 committed:
Updates to work with Polars 0.20.3 changes to is_in operator for categoricals

Believe due to Polars https://github.com/pola-rs/polars/pull/13205

mdahlhausen committed:
add required cbecs data file download

mdahlhausen committed:
botocore max connections increase

to prevent warning:
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: eulp.s3.us-west-2.amazonaws.com. Connection pool size: 20

mdahlhausen committed:
cast building type string as utf8 for polars

mdahlhausen committed:
plot if there are >=3 bldgs in ami instead of 5

mdahlhausen committed:
cast polars columns to correct dtype

mdahlhausen committed:
fix y max on day normalized ami comparison plots

mdahlhausen commented:
@asparke2:
- reran comparison script from new install of comstock repo and new environment
- added missing truth data
- cast dtypes for polars dataframes
- added README content
- fixed y max issue on day normalized plots
- fixed CBECS comparison test

Should be good to merge now.

I will add the week realizations and a summary of AMI comparisons for the documentation in a future PR
